cabby/geo/walk.py

```diff
@@ -38,56 +52,304 @@ def compute_route(start_point: Point, end_point: Point, graph: nx.MultiDiGraph,
+    '''Return a picked landmark on a given route.
+    Arguments:
+      route: Along the route a landmark will be chosen.
+      map: The map of a specific region.
+      end_point: The goal location.
+    Returns:
+      A single landmark.
+    '''
+
+    # Get POI along the route.
+    points_route = map.nodes[map.nodes['osmid'].isin(
+        route['osmid'])]['geometry']
+
+    try:
+        start_time = time.time()
+        poly = Polygon(points_route.tolist()).buffer(0.0005)
+        bounds = poly.bounds
+        bounding_box = box(bounds[0], bounds[1], bounds[2], bounds[3])
+        tags = {'wikipedia': True, 'wikidata': True, 'brand': True,
+                'tourism': True, 'amenity': True, 'shop': True, 'name': True}
+        df_pivots = ox.pois.pois_from_polygon(
+            bounding_box, tags=tags)
+
+    except Exception as e:
+        print(e)
+        return
+
+    # Polygon along the route.
+    df_pivots = df_pivots[df_pivots['geometry'].intersects(poly)]
+
+    # Remove streets.
+    df_pivots = df_pivots[(df_pivots['highway'].isnull())
+                          & (df_pivots['railway'].isnull())]
+
+    main_pivot = pick_prominent_pivot(df_pivots)
+    main_pivot = main_pivot.to_dict('records')[0]
+
+    # Get pivot near the goal location.
+    near_pivot = get_pivot_near_goal(map, end_point)
+    if main_pivot is None or near_pivot is None:
+        return
+
+    return main_pivot, near_pivot
+
+
+def get_points_and_route(map: map_structure.Map) -> Optional[Tuple[Dict, Dict, GeoDataFrame, Dict, Dict]]:
+    '''Sample start and end point, a pivot landmark and route.
+    Arguments:
+      map: The map of a specific region.
+    Returns:
+      A start and end point, a pivot landmark and route.
+    '''
+
+    # Select end point.
+    end_point = get_end_poi(map)
+    if end_point is None:
+        return
+
+    # Select start point.
+    start_point = get_start_poi(map, end_point)
+    if start_point is None:
+        return
+
+    # Compute route between start and end points.
+    route = compute_route(
+        start_point['centroid'], end_point['centroid'], map.nx_graph, map.nodes)
+    if route is None:
+        return
+
+    # Select pivots.
+    result = get_pivots(route, map, end_point)
+    if result is None:
+        return
+    main_pivot, near_pivot = result
+
+    return end_point, start_point, route, main_pivot, near_pivot
+
+
+def get_single_sample(map: map_structure.Map):
+    '''Sample start and end point, a pivot landmark and route and save to file.
+    Arguments:
+      map: The map of a specific region.
+    Returns:
+      A start and end point, a pivot landmark and route.
+    '''
+    result = get_points_and_route(map)
+    if result is None:
+        return
+    end_point, start_point, route, main_pivot, near_pivot = result
+    instruction = "Starting at {0} walk past {1} and your goal is {2}, near {3}.".format(
```

Comment:
I think it would be better to extract the geo info tuples and then do instruction generation condition on those, but we can leave this here for now.

Reply:
Okay, that's fine then.